### PR TITLE
Add material to stdio handling

### PR DESCRIPTION
**Pull request by mahkoh**

### CODE REVIEW

**Comment by mahkoh** [comment]
And here is how you would handle https://github.com/rust-lang/rfcs/pull/899#issuecomment-79091963 with this design:

Since you're writing a TUI, you're interacting with fds 0 and 1 directly anyway. Therefore, at the start of your program, you set `thread::set_stdout` to `io::Sink` and `thread::set_stderr` to a log file.

**Comment by l0kod** [comment]
Unfortunately, the Windows console do not make `Write` nor `Read` usable in a generic way.

cc #972

**Comment by mahkoh** [comment]
This RFC does not touch the raw versions.

**Comment by reem** [comment]
I might just not understand some fundamental limitation of the design space, but I'm curious why concurrent synchronization is being used (`Arc` and `Mutex`) to control a thread-local variable, which seems to only be modifiable from one thread. As a result, `Rc` and `RefCell` could be used, which would be more efficient.

**Comment by mahkoh** [comment]
@reem 

>I'm curious why concurrent synchronization is being used

```patch
+Unless other streams or another buffer mode has been explicitly specified via
+the thread-builder, these settings will be inherited by child-threads.
```

**Comment by reem** [comment]
Ah, I figured I had just missed something.

**Comment by aturon** [comment]
I think making `println` and `panic` reference a thread-local, inherited, overridable notion of `stdout` and `stderr` is a good move, for composability/capture purposes. So :+1: to the overall idea here.

One question: are the raw versions tied to the process-level handles? That is, does `std` give you any way to get at these underlying handles?

I'm a little unsure about the exact signatures/locations of some of the functions proposed here, but I want to think a bit more about a possible alternative for concrete comparison. I'll try to do so soon.

File: text/0517-io-os-reform.md [hunk 1]
```diff
@@ -1164,16 +1243,13 @@ pub fn stdout() -> Stdout;
...
-* `stdin` - returns a handle to a **globally shared** standard input of
-  the process which is buffered as well. Due to the globally shared nature of
-  this handle, all operations on `Stdin` directly will acquire a lock internally
-  to ensure access to the shared buffer is synchronized. This implementation
-  detail is also exposed through a `lock` method where the handle can be
-  explicitly locked for a period of time so relocking is not necessary.
+* `stdin` - wraps `thread::stdin`. All operations on `Stdin` directly
+  will acquire the lock internally. A `lock` method is provided to acquire
+  exclusive access to the stream for a period of time so that relocking is not
+  necessary.
```
Comment (liigo) [inline text/0517-io-os-reform.md]: Why lock on thread local streams?

Comment (mahkoh) [inline text/0517-io-os-reform.md]: They are not thread local.

**Comment by mahkoh** [comment]
@aturon 

>One question: are the raw versions tied to the process-level handles? That is, does std give you any way to get at these underlying handles?

This RFC does not make any changes to the raw section which is marked as a sketch anyway. Raw handles will have to refer to the 0, 1, 2 fds on unix directly.

**Comment by Ericson2314** [comment]
Is there any way to make some of this more opt-in so threads that don't do IO don't need to carry all this baggage around? All the behind the signs dynamic dispatch and thread local stuff reminds me a bit of the old green/native IO system.

I'm of the opinion that the standard streams are classic stupid arbitrary "singletons", and thus just plain bad design and holdovers from another era. The fact that we need to such things as are proposed here to fix them I take is evidence of the fact. I rather hook up panicking to the `liblog` infrastructure than bake this stuff into `std::thread`.

**Comment by mahkoh** [comment]
>all this baggage around?

Four pointers and a few atomic operations when you create a thread.

>classic stupid arbitrary "singletons"

You're suggesting removing thread local storage and println from the language. That's an interesting thought.

**Comment by aturon** [comment]
RFC issue https://github.com/rust-lang/rfcs/issues/1050

**Pull request closed without merging**